pkg/updatestatus/nodeinformer.go

```diff
@@ -202,160 +146,121 @@ func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncC
-	}
-
-	for k, v := range c.cache {
-		if k == mco.MachineConfigPoolMaster || k == mco.MachineConfigPoolWorker {
-			continue
+func (c *machineConfigPoolSelectorCache) whichMCP(l labels.Labels) string {
+	var ret string
+	c.cache.Range(func(k, v interface{}) bool {
+		s := v.(labels.Selector)
+		if k == mco.MachineConfigPoolMaster && s.Matches(l) {
+			ret = mco.MachineConfigPoolMaster
+			return false
 		}
-		if v.Matches(labels) {
-			return k
+		if s.Matches(l) {
+			ret = k.(string)
+			return ret == mco.MachineConfigPoolWorker
 		}
-	}
-
-	if v, ok := c.cache[mco.MachineConfigPoolWorker]; ok && v.Matches(labels) {
-		return mco.MachineConfigPoolWorker
-	}
-
-	return ""
+		return true
+	})
+	return ret
 }
 
-func (c *machineConfigPoolSelectorCache) ingest(pool *machineconfigv1.MachineConfigPool) bool {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
-	v, ok := c.cache[pool.Name]
+func (c *machineConfigPoolSelectorCache) ingest(pool *machineconfigv1.MachineConfigPool) (bool, string) {
 	s, err := metav1.LabelSelectorAsSelector(pool.Spec.NodeSelector)
 	if err != nil {
-		klog.Errorf("Failed to convert to a label selector from the node selector of MachineConfigPool %s : %v", pool.Name, err)
-		if ok {
-			delete(c.cache, pool.Name)
-			klog.V(4).Infof("Deleted MachineConfigPool %s from the cache", pool.Name)
-			return true
+		klog.Errorf("Failed to convert to a label selector from the node selector of MachineConfigPool %s: %v", pool.Name, err)
+		v, loaded := c.cache.LoadAndDelete(pool.Name)
+		if loaded {
+			return true, fmt.Sprintf("the previous selector %s for MachineConfigPool %s deleted as its current node selector cannot be converted to a label selector: %v", v, pool.Name, err)
 		} else {
-			return false
+			return false, ""
 		}
 	}
-	if !ok || v.String() != s.String() {
-		if c.cache == nil {
-			c.cache = make(map[string]labels.Selector)
+
+	previous, loaded := c.cache.Swap(pool.Name, s)
+	if !loaded || previous.(labels.Selector).String() != s.String() {
+		var vStr string
+		if loaded {
+			vStr = previous.(labels.Selector).String()
 		}
-		c.cache[pool.Name] = s
-		klog.V(4).Infof("Cached MachineConfigPool %s with selector %s", pool.Name, s.String())
-		return true
+		return true, fmt.Sprintf("selector for MachineConfigPool %s changed from %s to %s", pool.Name, vStr, s.String())
 	}
-	return false
+	return false, ""
 }
 
 func (c *machineConfigPoolSelectorCache) forget(mcpName string) bool {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	if _, ok := c.cache[mcpName]; ok {
-		delete(c.cache, mcpName)
-		klog.V(4).Infof("Deleted MachineConfigPool %s from the cache", mcpName)
-		return true
-	}
-	return false
+	_, loaded := c.cache.LoadAndDelete(mcpName)
+	return loaded
 }
 
-func (c *machineConfigPoolSelectorCache) len() int {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	return len(c.cache)
+func queueKeyFoReconcileAllNodes(queue workqueue.TypedRateLimitingInterface[any]) {
+	queue.Add(queueKeyFor(eventKindName, eventNameReconcileAllNodes))
```

Comment:
I think we should generally use AddRateLimited for robustness in error states

Reply:
:TIL
Done.